### PR TITLE
HPCC-11764 Dafilesrv authentication and encryption on transport

### PR DESCRIPTION
**Pull request by RussWhitehead**
Current dafilesrv uses unencrypted sockets which is not secure. This update
adds a new section to the environment.conf file to optionally specify that
dafilesrv and its clients should use SSL, and points to the cert file, key file
and provides an encrypted passphrase. If specified, dafilesrv and its clients
will enable secure sockets (openSSL) and communicate in a secure manner.

Signed-off-by: William Whitehead william.whitehead@lexisnexis.com

### CODE REVIEW

**Comment by jakesmith** [comment]
@RussWhitehead - can you look at my comments from 24th Nov.

File: system/jlib/jsocket.cpp [hunk 1]
```diff
@@ -6015,3 +6015,96 @@ ISocketConnectWait *nonBlockingConnect(SocketEndpoint &ep,unsigned connecttimeou
...
+        for (aindex_t idx = 0; res && socks.length(); idx++)
+        {
+            if (FD_ISSET(socks.item(idx), &fds))
+            {
+#ifdef _DEBUG
+                dbgSB.appendf(" %d",socks.item(idx));
+#endif
+                readySocks.append(socks.item(idx));
+                if (readySocks.length() == res)
+                    break;
```
Comment (mckellyln) [inline system/jlib/jsocket.cpp]: Probably never happen, but maybe put a max trip count on this loop with idx <= socks.length() ?

File: system/jlib/jutil.cpp [hunk 2]
```diff
@@ -2417,6 +2419,62 @@ IPropertyTree *getHPCCEnvironment(const char *configFileName)
...
+            try
+            {
+                StringBuffer configFileSpec;
+#ifndef _WIN32
+                configFileSpec.set(CONFIG_DIR).append(PATHSEPSTR).append("environment.conf");
+#endif
+                Owned<IProperties> conf = createProperties(configFileSpec.str(), true);
+                useSSL = conf->getPropBool("dfsUseSSL", false);
+                certificate.set(conf->queryProp("dfsSSLCertFile"));
+                privateKey.set(conf->queryProp("dfsSSLPrivateKeyFile"));
```
Comment (mckellyln) [inline system/jlib/jutil.cpp]: Should we always set certificate and privateKey if dfsUseSSL is false ?

File: common/remote/sockfile.cpp [hunk 3]
```diff
@@ -4365,12 +4431,67 @@ class CRemoteFileServer : public CInterface, implements IRemoteFileServer, imple
...
+            bool sockavail = false;
             try {
-                sockavail = acceptsock->wait_read(1000*60*1)!=0;
+                if (!useSSL)
+                    sockavail = acceptsock->wait_read(1000*60*1)!=0;
+                else
+                {
+                    UnsignedArray waitingSocks;
+                    //SSL Enabled. Listen for non SSL connection on DAFILESRV_PORT and reject them
+                    int numReady = wait_read_multiple(readSocks, 1000*60*1, waitingSocks);
```
Comment (mckellyln) [inline common/remote/sockfile.cpp]: I think using select() for 2 fds is fine.

File: system/jlib/jsocket.cpp (hunk 1 above)
Comment (RussWhitehead) [inline system/jlib/jsocket.cpp]: agreed, but i think should be    idx < socks.length()    ?

File: system/jlib/jutil.cpp (hunk 2 above)
Comment (RussWhitehead) [inline system/jlib/jutil.cpp]: I agree

File: common/remote/sockfile.cpp (hunk 3 above)
Comment (RussWhitehead) [inline common/remote/sockfile.cpp]: Leaving the code as is (using select() for 2 fds) unless someone protests

File: common/remote/sockfile.cpp [hunk 4]
```diff
@@ -4365,12 +4431,67 @@ class CRemoteFileServer : public CInterface, implements IRemoteFileServer, imple
...
+            querySecuritySettings(&useSSL, &configepPort, &clientCertificate.certificate, &clientCertificate.privateKey);
+            if (!clientCertificate.certificate)
+                throw createDafsException(DAFSERR_connection_failed,"SSL Certificate information not found in environment.conf");
+            if (listenep.port <= 0)
+                listenep.port = configepPort;
+
+            //Create unsecure socket to reject non-ssl client requests
+            StringBuffer ip;
+            IpAddress ipAddr;
+            acceptsock->getPeerAddress(ipAddr);
```
Comment (mckellyln) [inline common/remote/sockfile.cpp]: Are we connected to anything at this point ?
I get "getpeername failed 107" error msg

**Comment by mckellyln** [comment]
Note the passphrase must be removed from the key or else the program will fail with a bad password read error.  Is an uncrypted key file ok ?

**Comment by RussWhitehead** [comment]
Closing, and replacing with https://github.com/hpcc-systems/HPCC-Platform/pull/6921

**Pull request closed without merging**